add experimental gRPC implementation

mostynb opened the pull request:
This is a partial implementation of the Bazel Remote Execution API's
ActionCache, ContentAddressableStorage interfaces, as well as the
corresponding bytestream interface.

This is not ready for production use! I do not expect this pull request to land as-is, I'm mostly just looking for some early feedback.

Implements #64

mostynb commented:
Note: I haven't figured out how to build this with bazel yet (hence the buildkite failure), but linux-build.sh works for me.

mostynb commented:
Updated the PR to include bazel build support, but I had to resort to committing remote_execution.pb.go and semver.pb.go (with a couple of modified paths) to avoid dependency hell.

https://github.com/bazelbuild/remote-apis/blob/c0682f068a6044f395a7e28526abe1de56beffa8/build/bazel/remote/execution/v2/BUILD
depends on "googleapis" which is declared in
https://github.com/bazelbuild/remote-apis/blob/c0682f068a6044f395a7e28526abe1de56beffa8/WORKSPACE
as:
```
http_archive(
    name = "googleapis",
    build_file = "BUILD.googleapis",
    sha256 = "7b6ea252f0b8fb5cd722f45feb83e115b689909bbb6a393a873b6cbad4ceae1d",
    strip_prefix = "googleapis-143084a2624b6591ee1f9d23e7f5241856642f4d",
    urls = ["https://github.com/googleapis/googleapis/archive/143084a2624b6591ee1f9d23e7f5241856642f4d.zip"],
)
```
But that zip file does not contain a BUILD.googleapis file, nor any BUILD files with the required targets ("google_api_annotations_proto" et al). I suspect that "googleapis" is intended to be taken from this repository instead:
https://github.com/bazelbuild/bazel/blob/a8353612b4791531ae4e2ed463cf0e5d8a4009e8/third_party/googleapis/BUILD.bazel

I wonder if I should file an issue for this in https://github.com/bazelbuild/remote-apis ?

mostynb commented:
I found the magic incantation in github.com/bazelbuild/remote-apis-sdks, removed the .pb.go files.

mostynb commented:
Added GetTree and GetCapabilities.

buchgr commented:
Hi @mostynb thanks for the PR. I ll try to get to review it soon!

mostynb commented:
No problem. I'm still working on my test environment- do you know if it's possible to use bazel with remote caching via gRPC but not remote execution?

buchgr commented:
@mostynb yes it is possible and works the same was HTTP:  `--remote_cache=grpc://<host>:<port>`

mostynb commented:
Oh, nice- this allows me to do some real testing, and I immediately found a few bugs to fix. Pushed a new commit that:
* dropped an incorrect size check in GetActionResult. Aside: GetActionResultRequest.ActionDigest's SizeBytes field seems useless- do you think this is worth documenting in the .proto file?
* fixed an infinite loop in the bytestream Read function
* removed a noisy log

Now I am able to build bazel-remote using bazel with remote caching (via grpc), and then rebuild and get cache hits for everything except one ActionCache item at the end (I need to investigate to see why that is). I will also look for a other projects that use bazel for more testing.

mostynb commented:
> @mostynb As you are probably aware, you can compare your implementation to https://github.com/buildbarn/bb-storage/tree/master/pkg/cas.

I have seen that, though I haven't tried running it yet.

Right now I'm comparing bazel-remote using http and grpc, they both seem to give correct results on small bazel projects (like bazel-remote itself, and copybara). But I get quite low cache hit rates for larger porjects like kubernetes. I'm trying to figure out if this is non-deterministic/irreproducible builds on the kubernetes side, or a bug on the bazel-remote side.

mostynb commented:
Simply increasing my test cache size made the cache hit rate when building kubernetes go up to ~98%.

It turns out that the cached blobs are simply larger than I expected (maybe a golang thing?)- some larger than a gigabyte, and many larger than 10M.

mostynb commented:
Pushed an update with a bugfix in the bytestream Read function- the offset was being ignored, and when bazel received too much data back it would send back a new invalid Read request (offset larger than the size of the blob). With this fix, I can now build/rebuild large projects like kubernetes without errors.

My testcase:
bazel build --remote_cache=grpc://my.cache.server:9092 --remote_timeout=240 -- //... -//vendor/... -//build/rpms/...
The large timeout value seems to be required due to the large blobs mentioned previously.  I wonder if this is reporting as an issue to kubernetes (probably not worth caching >1G tarballs containing large go binaries).

Build performance is comparable to bb-storage, though I have not done any performance optimization yet.

TODO: read large blobs in chunks before sending them to the client, instead of buffering the entire thing.

mostynb commented:
Pushed an update which avoids buffering entire blobs in the bytestream implementation, and adds some tests.

mostynb commented on main.go:
```diff
@@ -157,6 +167,29 @@ func main() {
 		mux.HandleFunc("/", cacheHandler)
 
+		if c.GRPCPort > 0 {
+			go func() {
```
This initial implementation uses a separate port from the HTTP interface (for prototyping simplicity), which requires another blocking grpc Server.Serve call. If you prefer, I can try to change this to use the same port for both http and grpc and mux the requests and drop this extra goroutine.

buchgr replied:
Gottcha. I think it's absolutely fine for a single instance of bazel-remote to support either gRPC or http but not both. Do you have a use case where you want both?

mostynb replied:
Yes- I want to share a single large fast storage device between my goma backend which uses the HTTP interface and REAPI backends that use the gRPC interface, without a predefined amount allocated to each (which would be required with two bazel-remote instances). It's also useful for testing both implementations (eg with bazel which supports both).

With something like github.com/soheilhy/cmux we could use a single port for both (but there's still one goroutine for HTTP and one for gRPC in their examples).

mostynb commented on server/grpc_cas.go:
```diff
@@ -0,0 +1,248 @@
+
+// ContentAddressableStorageServer interface:
+
+func (s *grpcServer) FindMissingBlobs(ctx context.Context,
```
It's part of this interface in the protoc-generated code, so needs to exist:
https://github.com/bazelbuild/remote-apis/blob/4f46bfd2f8ab3354002610f0976f78ad2b68deba/build/bazel/remote/execution/v2/remote_execution.pb.go#L3304

I want bazel-remote to use usable by remote execution clients too, not just remote caching clients.

buchgr replied:
Makes sense. Thanks for clarifying.

mostynb commented on server/grpc_cas.go:
```diff
@@ -0,0 +1,248 @@
+	return &resp, nil
+}
+
+func (s *grpcServer) GetTree(in *pb.GetTreeRequest,
```
As with FindMissingBlobs, it's part of the generated ContentAddressableStorageServer interface, and needs to exist.

IIRC I initially tested this with the remote-apis-sdks cas client.

mostynb commented on server/grpc_ac.go:
```diff
@@ -0,0 +1,127 @@
+		return nil, status.Error(codes.Unknown, err.Error())
+	}
+
+	result := &pb.ActionResult{}
```
Adding a check of this type on the client side fixed the most interesting bug I encountered when implementing our goma backend...

Unfortunately, implementing this check for the HTTP interface would break my usage of bazel-remote with our custom goma backend- we store arbitrary blobs using /ac/<key>. I could add such a check if we added a new REST endpoint which skipped these checks, eg /raw/<key> - what do you think? Then I could just rename the existing "ac" directory to "raw" in our cache and not have to start over with an empty cache.

buchgr replied:
Sure. I think adding the check to HTTP should be done in a separate PR. We can discuss then.

mostynb commented on server/grpc_ac.go:
```diff
@@ -0,0 +1,127 @@
+
+	s.accessLogger.Printf("GRPC AC PUT %s OK", req.ActionDigest.Hash)
+
+	// Why do we need to return the ActionResult from the input?
```
Replaced this comment with a pointer to the style guide.

mostynb commented on server/grpc_bytestream.go:
```diff
@@ -0,0 +1,361 @@
+const (
+	// The maximum chunk size to write back to the client in Send calls.
+	// TODO: decide on a final value, or make this configurable?
+	// Inspiration:
```
WFM, removed the comments about the other values.

mostynb commented on server/grpc_bytestream.go:
```diff
@@ -0,0 +1,361 @@
+
+func (s *grpcServer) Write(srv bytestream.ByteStream_WriteServer) error {
+
+	req, err := srv.Recv()
```
I used two Recv calls because there's some code that should only run after the first call. I could use a single Recv call in a for loop with an if block for the first iteration, if you prefer.

buchgr replied:
> I could use a single Recv call in a for loop with an if block for the first iteration, if you prefer.

That would be great. I think this would help readability a lot! Thanks!

mostynb replied:
Pushed a fixup with this change.

mostynb commented on server/grpc_bytestream.go:
```diff
@@ -0,0 +1,361 @@
+			"Negative ReadOffset is out of range")
+	}
+
+	// TODO: take req.ReadLimit into account?
```
I'm not sure if you're referring to req.ReadOffset or req.ReadLimit, but IIUC both refer to a position in the "resource", while the blob size is an arbitrary server implementation detail.

I added some code in a fixup to resolve the req.ReadLimit TODO.

mostynb commented on server/grpc_bytestream.go:
```diff
@@ -0,0 +1,361 @@
+	return nil
+}
+
+func (s *grpcServer) QueryWriteStatus(context.Context, *bytestream.QueryWriteStatusRequest) (*bytestream.QueryWriteStatusResponse, error) {
```
I meant that this can't operate on a partial/failed bs.Write call with bazel-remote specifically, because bazel-remote does not support partial/failed CAS writes (as far as I can tell it just deletes the data and the client has to start from the beginning).

Anyway updated to return 0, which I guess is OK for bazel and I suspect no other clients use this API.

mostynb commented on server/grpc_bytestream.go:
```diff
@@ -0,0 +1,361 @@
+
+	buf := make([]byte, bufSize)
+
+	if req.ReadOffset > 0 {
```
Isn't QueryWriteStatus for bs.Write calls, whereas this is a bs.Read operation?

buchgr commented:
Thanks @mostynb !
